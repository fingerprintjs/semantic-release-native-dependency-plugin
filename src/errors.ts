export class JsonParseError extends Error {
  constructor(podspecFilePath: string) {
    super(`${podspecFilePath} file cannot be parsed as JSON.`)
  }
}
