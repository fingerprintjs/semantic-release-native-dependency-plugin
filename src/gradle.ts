import { join } from 'node:path'
import { access, constants } from 'node:fs'
import { exec } from 'node:child_process'

/**
 * Check if gradle is installed to system-wide.
 * @return A promise that resolves if gradle is found
 */
export function isGradleAvailable(): Promise<boolean> {
  return new Promise((resolve, reject) => {
    exec('gradle --version', (err) => {
      if (err) {
        // 127 means command not found (https://tldp.org/LDP/abs/html/exitcodes.html)
        if (err.code === 127) {
          resolve(false)
        }
        reject(err)
      }

      resolve(true)
    })
  })
}

/**
 * @param {string} cwd the path of current working directory
 * @return A promise that resolves name of command to trigger Gradle
 */
export function getCommand(cwd: string): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    const gradleWrapper = join(cwd, 'gradlew')
    access(gradleWrapper, constants.F_OK, async (err) => {
      if (err) {
        if (err.code === 'ENOENT') {
          if (await isGradleAvailable()) {
            resolve('gradle')
          }

          reject(new Error('Gradle or Gradle wrapper not found.'))
        } else {
          reject(err)
        }
      } else {
        resolve(gradleWrapper)
      }
    })
  })
}
