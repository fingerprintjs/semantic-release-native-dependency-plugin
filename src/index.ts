export { default as generateNotes } from './generateNotes'
