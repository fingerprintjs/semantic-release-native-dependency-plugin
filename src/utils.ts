export function humanizeMavenStyleVersionRange(versionRange: string) {
  return versionRange
    .trim()
    .replace(/\s*\[\s*(.*?)\s*,\s*(.*?)\s*]\s*/g, '>= $1 and <= $2')
    .replace(/\s*\(\s*(.*?)\s*,\s*(.*?)\s*]\s*/g, '> $1 and <= $2')
    .replace(/\s*\[\s*(.*?)\s*,\s*(.*?)\s*\)\s*/g, '>= $1 and < $2')
    .replace(/\s*\(\s*(.*?)\s*,\s*(.*?)\s*\)\s*/g, '> $1 and < $2')
    .replace(/\s*\[\s*(.*?)\s*]/g, '>=$1')
    .replace(/\s*\(\s*(.*?)\s*\)/g, '>$1')
}

export function isNotFoundErrorCode(code?: number) {
  // 127 means command not found (https://tldp.org/LDP/abs/html/exitcodes.html)
  return code === 127
}
