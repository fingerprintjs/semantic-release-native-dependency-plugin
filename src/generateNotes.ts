import { join } from 'node:path'
import type { GenerateNotesContext } from 'semantic-release'
import { spawn } from 'node:child_process'
import type { Signale } from 'signale'
import split from 'split2'
import PluginConfig from './@types/pluginConfig'
import { getCommand } from './gradle'
import { humanizeMavenStyleVersionRange } from './utils'
import { readFileSync } from 'node:fs'

export async function getAndroidVersion(
  cwd: string,
  androidPath: string,
  androidGradleTaskName: string,
  env: NodeJS.ProcessEnv,
  logger: Signale
): Promise<string> {
  const androidFullPath = join(cwd, androidPath)
  const command = await getCommand(androidFullPath)

  return new Promise((resolve, reject) => {
    const child = spawn(command, [androidGradleTaskName, '-q', '--console=plain'], {
      cwd: androidPath,
      env,
      detached: true,
      stdio: ['inherit', 'pipe', 'pipe'],
    })
    if (child.stdout === null) {
      reject(new Error('Unexpected error: stdout of subprocess is null'))
      return
    }
    if (child.stderr === null) {
      reject(new Error('Unexpected error: stderr of subprocess is null'))
      return
    }

    let androidVersion: string | null = null
    child.stdout.pipe(split()).on('data', (line: string) => {
      androidVersion = line
    })
    child.stderr.pipe(split()).on('data', (line: string) => {
      logger.error(line)
    })
    child.on('close', (code: number) => {
      if (code !== 0) {
        reject(new Error(`Unexpected error: Gradle failed with status code ${code}`))
        return
      }

      if (androidVersion === null) {
        reject(new Error('Could not read output of `printFingerprintNativeSDKVersion` gradle task.'))
        return
      }

      resolve(androidVersion)
    })
    child.on('error', (err) => {
      logger.error(err)
      reject(err)
    })
  })
}

type PodspecJson = {
  dependencies: {
    [key: string]: [string]
  }
}

export const getIOSVersion = async (cwd: string, iOSPodSpecJsonPath: string) => {
  const jsonFile = join(cwd, iOSPodSpecJsonPath)

  let fileContent: string
  try {
    fileContent = readFileSync(jsonFile, 'utf8')
  } catch (error: any) {
    switch (error.code) {
      case 'ENOENT':
        throw new Error(`${iOSPodSpecJsonPath} file does not exist.`)
      case 'EACCES':
        throw new Error(`${iOSPodSpecJsonPath} file cannot be accessed.`)
      default:
        throw new Error(`${iOSPodSpecJsonPath} file cannot be read.`)
    }
  }

  let data: PodspecJson
  try {
    data = JSON.parse(fileContent) as PodspecJson
  } catch (error) {
    throw new Error(`${iOSPodSpecJsonPath} file cannot be parsed as JSON.`)
  }

  if (!data.dependencies || !data.dependencies['FingerprintPro']) {
    throw new Error(`${iOSPodSpecJsonPath} file does not contain 'FingerprintPro' in dependencies.`)
  }

  return data.dependencies['FingerprintPro'].join(' and ')
}

const generateNotes = async (
  { androidPath, androidGradleTaskName, iOSPodSpecJsonPath }: PluginConfig,
  { logger, cwd, env }: GenerateNotesContext
) => {
  if (!cwd) {
    throw new Error(`Current working directory is required to detect native SDK versions.`)
  }

  if (!androidGradleTaskName) {
    throw new Error('Android gradle task name should be defined.')
  }

  const androidVersion = await getAndroidVersion(
    cwd,
    androidPath,
    androidGradleTaskName,
    env as NodeJS.ProcessEnv,
    logger
  )
  const humanizedAndroidVersion = humanizeMavenStyleVersionRange(androidVersion)
  logger.log(`Detected Android Native SDK Version: \`${androidVersion}\` \`${humanizedAndroidVersion}\``)

  if (!iOSPodSpecJsonPath) {
    throw new Error('iOS Podspec Json path should be defined.')
  }

  const iosVersion = await getIOSVersion(cwd, iOSPodSpecJsonPath)
  logger.log(`Detected iOS Native SDK Version: \`${iosVersion}\``)

  return `Fingerprint Android SDK Version Range: **\`${humanizedAndroidVersion}\`**

Fingerprint iOS SDK Version Range: **\`${iosVersion}\`**`
}

export default generateNotes
