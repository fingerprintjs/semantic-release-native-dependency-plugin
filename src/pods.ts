import { exec } from 'child_process'
import { isNotFoundErrorCode } from './utils'
import { JsonParseError } from './errors'

/**
 * Checks whether CocoaPods are installed and available on the system by executing the `pod --version` command.
 * Resolves to `true` if CocoaPods are available, resolves to `false` if CocoaPods are not installed,
 * and rejects with an error if another issue occurs.
 *
 * @return {Promise<boolean>} A promise that resolves to `true` if CocoaPods are installed `false` otherwise.
 */
export function isPodAvailable(): Promise<boolean> {
  return new Promise((resolve, reject) => {
    exec('pod --version', (err) => {
      if (err) {
        if (isNotFoundErrorCode(err.code)) {
          resolve(false)
        }
        reject(err)
      }

      resolve(true)
    })
  })
}

/**
 * Executes the `pod ipc spec` command on the given podspec file and returns the parsed result as an object.
 *
 * @param {string} podspecFile - The path to the `.podspec` file to be processed.
 * @return {Promise<PodspecJson>} A promise that resolves to the parsed PodspecJson object or rejects with an error if the command fails.
 */
export function podIpcSpec(podspecFile: string): Promise<PodspecJson> {
  return new Promise<PodspecJson>((resolve, reject) => {
    exec(`pod ipc spec ${podspecFile}`, (err, stdout) => {
      if (err) {
        reject(err)
      }

      try {
        resolve(JSON.parse(stdout) as PodspecJson)
      } catch {
        throw new JsonParseError(podspecFile)
      }
    })
  })
}

export type PodspecJson = {
  dependencies: {
    [key: string]: [string]
  }
}
