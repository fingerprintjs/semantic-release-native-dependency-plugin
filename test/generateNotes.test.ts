import { describe, expect, it } from '@jest/globals'
import { getAndroidVersion, getIOSVersion } from '../src/generateNotes'
import { cwd } from 'node:process'
import { Signale } from 'signale'

describe('Test for generating notes', () => {
  describe('getAndroidVersion function', () => {
    it('returns correct version', async () => {
      const command = await getAndroidVersion(
        cwd(),
        'test/project/with-wrapper',
        'printFingerprintNativeSDKVersion',
        process.env,
        new Signale()
      )
      expect(command).toBe('[1.2.3, 4.5.6)')
    })
    it('throws error when task not found', () => {
      const signale = new Signale()
      signale.disable()

      expect(
        getAndroidVersion(cwd(), 'test/project/with-wrapper', 'nonExists', process.env, signale)
      ).rejects.toThrowErrorMatchingSnapshot('gradleNoTask')
    })
  })
  describe('getIOSVersion function', () => {
    it('returns correct version', async () => {
      const iosVersion = await getIOSVersion(cwd(), 'test/project/ios/podspec.json')
      expect(iosVersion).toBe('>= 1.2.3 and < 4.5.6')
    })
    it('throws error when file not found', () => {
      expect(getIOSVersion(cwd(), 'nonExists')).rejects.toThrowErrorMatchingSnapshot('podspecFileNotFound')
    })
  })
})
