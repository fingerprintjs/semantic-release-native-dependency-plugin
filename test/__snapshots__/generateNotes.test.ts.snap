// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for generating notes getIOSVersion function throws error when file not found: podspecFileNotFound 1`] = `"ENOENT: no such file or directory, open '/Users/eray/Code/Company/Fingerprint/Projects/semantic-release-native-sdk-plugin/nonExists'"`;
